### PR TITLE
Event bus

### LINKED ISSUE #455: Implement EventBus
## User stories
As a developer I want to be able to subscribe to events happening in the simuation in order to react to them

## Depends on
- #443 

## (Wiki) References
- [EventBus](https://github.com/BP2022-AP1/bp2022-ap1/wiki/EventBus)

## Acceptance Criteria
- [ ] Implement `EventConsumer` that can subscribe und unsubscribe to events with callbacks
- [ ] Implement `EventBus` that replaces the logger in all components and takes all event notifications
- [ ] Make the `Logger` an `EventConsumer`
- [ ] Implement `Event` and  `EventType` to use as argument for event callbacks

### PR DESCRIPTION
**Pull request by instant-sky**
Fixes #455 

Adds an event bus that enables components to react to events.
For documentation see https://github.com/BP2022-AP1/bp2022-ap1/wiki/EventBus#updated-version-of-more-indirection

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [x] Breaking changes anounced
- [ ] ~~Added classes that inherit from `BaseModel` to `src.constants.tables`~~
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

File: src/event_bus/event_bus.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        """Cancels the subscription associated to a handle
+
+        :param handle: the handle
+        """
+        self.callbacks.pop(handle)
+
+    def _get_callbacks_for_event_type(
+        self, event_type: EventType
+    ) -> Iterator[Callable[[Event], None]]:
+        """Returns an iterator to all callbacks associated with th eenevent type
```
Comment (Lietze) [inline src/event_bus/event_bus.py]: Typo
```suggestion
        """Returns an iterator to all callbacks associated with the relevant type
```

File: src/logger/logger.py [hunk 2]
```diff
@@ -21,470 +24,521 @@
...
-        :param tick: The current simulation tick
-        :param train_speed_fault_configuration: The configuration of the fault
-        :param affected_element: The affected element
-        :param value_before: The value before the fault
-        :param value_after: The value after the fault
-        :rtype: None
+        :param event: the event containing all relevant info
         """
+        if "train_speed_fault_configuration" not in event.arguments:
+            return
```
Comment (Lietze) [inline src/logger/logger.py:412]: Why did you add this? I think that should not ever happen right?

File: src/logger/logger.py [hunk 3]
```diff
@@ -21,470 +24,521 @@
...
         """
         This function should be called when removing a platform blocked fault from the simulation.
         This should reference the fault configuration of the fault.
-        :param tick: The current simulation tick
-        :param platform_blocked_fault_configuration: The configuration of the fault
-        :rtype: None
+        :param event: the event containing all relevant info
         """
+        if "platform_blocked_fault_configuration" not in event.arguments:
+            return
```
Comment (Lietze) [inline src/logger/logger.py:436]: Same here. Does this ever happen?

Comment (Lietze) [reply src/logger/logger.py:436]: Also everywhere else, I'm not writing the same comment 20 times ;)

File: src/event_bus/event_bus.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        defined in EVENT_METHODS. It returns a callable that creates
+        an event and calls all callbacks.
+
+        :param name: the method name
+        :raises AttributeError: when unknown method name is passed
+        :return: a callable
+        """
+        if name not in self.EVENT_METHODS:
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{name}'"
```
Comment (Lietze) [inline src/event_bus/event_bus.py:197]: Why does self. __class__. __name__ not always refer to EventBus? If it does, I think you should make it explicit.

File: src/event_bus/event_bus.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            )
+        arg_keys, event_type = self.EVENT_METHODS[name]
+
+        def emit_event(*args: list, **kwargs: dict):
+            arguments = self._build_arguments(name, arg_keys, args, kwargs)
+            event = Event(event_type, arguments)
+            for callback in self._get_callbacks_for_event_type(event_type):
+                callback(event)
+
+        return emit_event
```
Comment (Lietze) [inline src/event_bus/event_bus.py:207]: How exacly does __getattr__ work and when is it called?
Is it called whenever any method or attribute of this class is called to find it and you are overwriting this?
And why does the return value get executed in the other components?

File: src/logger/logger.py [hunk 6]
```diff
@@ -21,470 +24,521 @@
...
     run_id: UUID
+    callback_handles: list[UUID]
 
-    def __init__(self, run_id: UUID):
+    def __init__(self, run_id: UUID, event_bus: EventBus):
         """
         The constructor of the logger class
         """
+        super().__init__(event_bus, 5)
         self.run_id = run_id
```
Comment (Lietze) [inline src/logger/logger.py]: Does the Logger still need run_id for anything or could it be removed? I was wondering, because the EventBus also has run_id.

File: tests/fault_injector/conftest.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import pytest
 from traci import vehicle
 
+from src.event_bus.event_bus import EventBus
 from src.implementor.models import Run, SimulationConfiguration, Token
 from src.interlocking_component.route_controller import IInterlockingDisruptor
 from src.logger.logger import Logger
```
Comment (Lietze) [inline tests/fault_injector/conftest.py:7]: How come you still need the Logger?

File: tests/fault_injector/test_affected_element_does_not_exist.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 import pytest
 
 from src.base_model import BaseModel
+from src.event_bus.event_bus import EventBus
 from src.fault_injector.fault_configurations.fault_configuration import (
     FaultConfiguration,
 )
```
Comment (Lietze) [inline tests/fault_injector/test_affected_element_does_not_exist.py:7]: Further down there is still the Logger as an import. Do you still need that?

File: tests/fault_injector/test_platform_blocked_fault.py [hunk 9]
```diff
@@ -1,5 +1,6 @@
 import pytest
 
+from src.event_bus.event_bus import EventBus
 from src.fault_injector.fault_configurations.platform_blocked_fault_configuration import (
     PlatformBlockedFaultConfiguration,
 )
```
Comment (Lietze) [inline tests/fault_injector/test_platform_blocked_fault.py:6]: Logger is still there as an import.

File: tests/fault_injector/test_schedule_blocked_fault.py [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import pytest
 
+from src.event_bus.event_bus import EventBus
 from src.fault_injector.fault_configurations.schedule_blocked_fault_configuration import (
     ScheduleBlockedFaultConfiguration,
 )
```
Comment (Lietze) [inline tests/fault_injector/test_schedule_blocked_fault.py:6]: Logger as Import

File: tests/fixtures/fixtures_spawner.py [hunk 11]
```diff
@@ -757,15 +759,22 @@ def let_next_spawn_fail(self):
     return MockTrainSpawner()
 
 
+@pytest.fixture
+def event_bus(run) -> EventBus:
+    bus = EventBus(run_id=run.id)
+    Logger(run.id, bus)
+    return bus
```
Comment (Lietze) [inline tests/fixtures/fixtures_spawner.py]: This is the same fixture as in fixtures/fixtures_logger.py. Why do you need it again?

File: tests/interlocking_component/conftest.py [hunk 12]
```diff
@@ -137,10 +137,10 @@ def get_gg_count() -> int:
 
 @pytest.fixture
 def route_controller(
-    configured_souc: SimulationObjectUpdatingComponent, mock_logger: Logger
+    configured_souc: SimulationObjectUpdatingComponent, event_bus: EventBus
```
Comment (Lietze) [inline tests/interlocking_component/conftest.py]: Shouldn't this be different?
```suggestion
    configured_souc: SimulationObjectUpdatingComponent, mock_event_bus: EventBus
```

Comment (Lietze) [reply tests/interlocking_component/conftest.py]: Also apply any changes two lines down.

File: tests/logger/test_log_collector.py [hunk 13]
```diff
@@ -224,342 +225,348 @@ def _enter_leave_block_section_all_df(self):
...
+        self.setup_departure_arrival_1(event_bus)
         assert_frame_equal(
-            log_collector.get_departures_arrivals_all_trains(logger.run_id),
+            log_collector.get_departures_arrivals_all_trains(event_bus.run_id),
             _departure_arrival_all_1_df,
         )
 
-    def test_departure_arrival_empty(self, logger: Logger, log_collector: LogCollector):
+    def test_departure_arrival_empty(
+        self, event_bus: Logger, log_collector: LogCollector
```
Comment (Lietze) [inline tests/logger/test_log_collector.py]: Why is it `event_bus: Logger`? Does this make the typehints better or is there a real reason?

File: tests/logger/test_log_collector.py [hunk 14]
```diff
@@ -581,5 +588,7 @@ def test_get_faults(
         faults_log_collector_df["fault_id"] = faults_log_collector_df[
             "fault_id"
         ].astype("string")
-        generated_faults_df = log_collector.get_faults(logger.run_id)
+        generated_faults_df = log_collector.get_faults(event_bus.run_id)
+        print(generated_faults_df)
+        print(faults_log_collector_df)
```
Comment (Lietze) [inline tests/logger/test_log_collector.py]: Please remove these two lines.
```suggestion
```

File: src/logger/logger.py [hunk 15]
```diff
@@ -21,470 +24,521 @@
...
+        :param event: the event containing all relevant info
         """
         TrainLeaveBlockSectionLogEntry.create(
             timestamp=datetime.now(),
-            tick=tick,
-            message=f"Train with ID {train_id} left block section with ID {block_section_id}",
+            tick=event.arguments["tick"],
+            message=f"Train with ID {event.arguments['train_id']} left block "
+            f"section with ID {event.arguments['block_section_id']} "
+            f"with length {event.arguments['block_section_length']}",
```
Comment (Lietze) [inline src/logger/logger.py]: Why did you add to the message?

File: tests/event_bus/test_event_bus.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    def test_event(
+        self, event_bus: EventBus, event_type: EventType, tick: int, train_id: str
+    ):
+        self.callback_called = False
+
+        def callback(event: Event):
+            self.callback_called = True
+            assert event.event_type == event_type
+            assert event.arguments["tick"] == tick
+            assert event.arguments["train_id"] == train_id
```
Comment (Lietze) [inline tests/event_bus/test_event_bus.py:65]: Is there an way to make the asserts outside of the callback? I think that would be easier to understand.

**Review by Lietze — CHANGES_REQUESTED** [review]
This looks great. I have a few questions and I fould a few things you may have overlooked.

File: src/logger/logger.py (hunk 2 above)
Comment (instant-sky) [reply src/logger/logger.py:412]: This can be the case, because there are only the EventTypes `INJECT_FAULT` and `RESOLVE_FAULT`. This results in all of the specific faults have to check, if they are meant by the call or it addresses another fault type.

File: src/logger/logger.py (hunk 3 above)
Comment (instant-sky) [reply src/logger/logger.py:436]: see above

File: src/event_bus/event_bus.py (hunk 4 above)
Comment (CR1337) [reply src/event_bus/event_bus.py:197]: This version minimizes the amount of places where you would have to change code when you would rename the class. It therefore reduces redundancy. But if it really bugs you I can change it and rely on modern IDE magic for renaming :)

File: src/event_bus/event_bus.py (hunk 5 above)
Comment (CR1337) [reply src/event_bus/event_bus.py:207]: `__getattr__` gets called when the caller tries to acces an attribute that doesn't exist. It then creates this attribute and returns it. Because the caller here tries to execute a method, `__getattr__` has to return a callable that then gets executed by the caller.

File: tests/logger/test_log_collector.py (hunk 13 above)
Comment (CR1337) [reply tests/logger/test_log_collector.py]: We just overlooked this. It will be changed :)

File: tests/event_bus/test_event_bus.py (hunk 16 above)
Comment (CR1337) [reply tests/event_bus/test_event_bus.py:65]: I don't know how. I want to test if the argument passed to the callback contains the correct data. 

Do you have an idea how to test it otside the callback?

File: src/event_bus/event_bus.py (hunk 4 above)
Comment (Lietze) [reply src/event_bus/event_bus.py:197]: No this makes sense to me. God job

File: src/event_bus/event_bus.py (hunk 5 above)
Comment (Lietze) [reply src/event_bus/event_bus.py:207]: All right. This feels very efficient but also very complicated at the same time. I think it is a good way of doing it, but maybe you could add some comments to prevent some headaches.

File: tests/event_bus/test_event_bus.py (hunk 16 above)
Comment (Lietze) [reply tests/event_bus/test_event_bus.py:65]: No. This is also fine.

File: src/event_bus/event_bus.py (hunk 5 above)
Comment (CR1337) [reply src/event_bus/event_bus.py:207]: I didn't want to create a method for every possible event which would effectively all do the same.

But you are right. I will add some comments to clarify the functionality.

File: src/logger/logger.py (hunk 6 above)
Comment (CR1337) [reply src/logger/logger.py]: You are right. I removed it. Thanks :)

File: tests/fault_injector/conftest.py (hunk 7 above)
Comment (CR1337) [reply tests/fault_injector/conftest.py:7]: Prior to this refactoring tests using fixtures from this file used the `Logger` and expected `LogEntry`s to be created. Now `Components` have the `EventBus` instead of the `Logger`. Nonetheless `LogEntry`s still need to be created. Therefore the Logger still needs to be instantiated here.

File: tests/fault_injector/test_affected_element_does_not_exist.py (hunk 8 above)
Comment (CR1337) [reply tests/fault_injector/test_affected_element_does_not_exist.py:7]: I will remove it.

File: tests/fault_injector/test_platform_blocked_fault.py (hunk 9 above)
Comment (CR1337) [reply tests/fault_injector/test_platform_blocked_fault.py:6]: Will be removed.

File: tests/fault_injector/test_schedule_blocked_fault.py (hunk 10 above)
Comment (CR1337) [reply tests/fault_injector/test_schedule_blocked_fault.py:6]: Will be removed.

File: tests/fixtures/fixtures_spawner.py (hunk 11 above)
Comment (CR1337) [reply tests/fixtures/fixtures_spawner.py]: There was already some duplication prior to this refactoring. The duplicate fixtures partly rely on different other fixtures. If I simply remove the duplicate fixtures it breaks a number of tests. This should definitely be taken care of but I suggest doing it in a future issue due to the complexity this problem seems to have.

File: src/logger/logger.py (hunk 15 above)
Comment (CR1337) [reply src/logger/logger.py]: I will change it back. The length is stored in the `LogEntry` anyway.

File: tests/data_science/test_data_science.py [hunk 17]
```diff
@@ -136,70 +141,66 @@ def test_get_coal_demand_by_run_id(
...
-        data_science: DataScience,
-        window_size_time_by_config_id_df: pd.DataFrame,
-    ):
-        setup_logs_departure_arrival(logger)
-        setup_logs_departure_arrival_alt(logger2)
-        window_size_df = data_science.get_window_size_time_by_config_id(
-            simulation_configuration
-        )
-        assert_frame_equal(window_size_df, window_size_time_by_config_id_df)
-
```
Comment (Lietze) [inline tests/data_science/test_data_science.py:194]: Why did you just remove this test?

File: tests/data_science/test_data_science.py [hunk 18]
```diff
@@ -255,13 +256,13 @@ def test_get_coal_spawn_events_by_config_id(
     def test_get_window_by_config_id(
         self,
         simulation_configuration: SimulationConfiguration,
-        logger: Logger,
-        logger2: Logger,
+        event_bus: EventBus,
+        event_bus2: Logger,
```
Comment (Lietze) [inline tests/data_science/test_data_science.py]: Why is this a Logger?

**Review by Lietze — COMMENTED** [review]
I was wondering why you deleted the test from the data_science module? Also I think in a lot of tests there is still a Logger being used when it should be a EventBus. You should try to catch all of those.

File: tests/fixtures/fixtures_spawner.py (hunk 11 above)
Comment (Lietze) [reply tests/fixtures/fixtures_spawner.py]: That seems good.

File: tests/data_science/test_data_science.py (hunk 17 above)
Comment (CR1337) [reply tests/data_science/test_data_science.py:194]: The method that was tested was removed during development on this branch. After merging dev into here I had to delete this test manually because GitHub wasn't able to.

File: tests/data_science/test_data_science.py (hunk 18 above)
Comment (CR1337) [reply tests/data_science/test_data_science.py]: The `logger` here is passed to `setup_logs_departure_arrival` and `setup_logs_departure_arrival_alt`. Both prepare the `TestLogCollector` which directly interacts with the Logger. Therefore it needs a `Logger` to be passed as argument.

File: tests/data_science/test_data_science.py (hunk 17 above)
Comment (Lietze) [reply tests/data_science/test_data_science.py:194]: `get_window_size_time_by_config_id` does still exist in data_science.py in line 397. Or did you mean something else?

File: tests/data_science/test_data_science.py (hunk 18 above)
Comment (Lietze) [reply tests/data_science/test_data_science.py]: If `setup_logs_departure_arrival` and `setup_logs_departure_arrival_alt` both prepare the TestLogCollector, why does one get a Logger and the other a EventBus? If the fixture of the EventBus is passed, then it should just execute the callback to the logger anyway right?
What I'm not sure about is, if the data_science fixture would still have access to the created logs.

**Review by Lietze — APPROVED** [review]
Very nice, thank you!

File: tests/data_science/test_data_science.py (hunk 18 above)
Comment (CR1337) [reply tests/data_science/test_data_science.py]: I changed the type hints to `EventBus` now.

File: tests/data_science/test_data_science.py (hunk 17 above)
Comment (CR1337) [reply tests/data_science/test_data_science.py:194]: I don't know what went wrong here. I reenabled the test.
